donkeycar/parts/tub_v2.py

```diff
@@ -109,24 +117,79 @@ class TubWriter(object):
     """
     A Donkey part, which can write records to the datastore.
     """
+
+    class TubRecordCached:
+        """
+        Simple helper for storeing record in a queue
+        """
+        def __init__(self, record, timestamp):
+            self.record = record
+            self.timestamp = timestamp
+
     def __init__(self, base_path, inputs=[], types=[], metadata=[],
                  max_catalog_len=1000):
         self.tub = Tub(base_path, inputs, types, metadata, max_catalog_len)
+        self.record_queue = Queue()
+        self.mutex = Lock()
+        self.on = False
 
     def run(self, *args):
         assert len(self.tub.inputs) == len(args), \
             f'Expected {len(self.tub.inputs)} inputs but received {len(args)}'
         record = dict(zip(self.tub.inputs, args))
-        self.tub.write_record(record)
+        self.tub.write_record(record, time.time())
         return self.tub.manifest.current_index
 
-    def __iter__(self):
-        return self.tub.__iter__()
+    def run_threaded(self, *args):
+        assert len(self.tub.inputs) == len(args), \
+            f'Expected {len(self.tub.inputs)} inputs but received {len(args)}'
+        record = dict(zip(self.tub.inputs, args))
+        recordCached = self.TubRecordCached(record, time.time())
+        # push the cached record into queue
```

Comment:
TubRecordCached is not an instance variable. Also why are we introducing a new class for a cached TubRecord ?

Reply:
It's not an instance variable. TubRecordCached is a nested class. We need it because we want to cache the record in the queue, along with the timestamp at that time. We can't use `time.time()` in the writer thread as the timestamp will not be accurate. Thus we need this little class to do the job.

Reply:
We don't need to do that. The timestamp does not have to be precise. Also, it does not have to be nested either way.

Reply:
Well I'm not an expert what timestamp is to do here but if there's a way to record it as precisely as before for both threaded and non-threaded mode at no cost then why not? I made it nested because I don't want to pollute the global namespace so it's just a little implementation details inside `TubWriter`.